include/PluginPinConnector.h

```diff
@@ -142,22 +198,539 @@ class LMMS_EXPORT PluginPinConnector
+
+		for (std::uint8_t inChannelPairIdx = 0; inChannelPairIdx < inSizeConstrained; ++inChannelPairIdx)
+		{
+			const SampleFrame* inPtr = in.bus[inChannelPairIdx]; // L/R track channel pair
+
+			const std::uint8_t inChannel = inChannelPairIdx * 2;
+			const std::uint8_t enabledPins =
+				(static_cast<std::uint8_t>(m_pc->m_in.enabled(inChannel, outChannel)) << 1u)
+				| static_cast<std::uint8_t>(m_pc->m_in.enabled(inChannel + 1, outChannel));
+
+			switch (enabledPins)
+			{
+				case 0b00: break;
+				case 0b01: // R channel only
+				{
+					for (f_cnt_t frame = 0; frame < in.frames; ++frame)
+					{
+						outPtr[frame] += convertSample<SampleT>(inPtr[frame].right());
+					}
+					++numRouted;
+					break;
+				}
+				case 0b10: // L channel only
+				{
+					for (f_cnt_t frame = 0; frame < in.frames; ++frame)
+					{
+						outPtr[frame] += convertSample<SampleT>(inPtr[frame].left());
+					}
+					++numRouted;
+					break;
+				}
+				case 0b11: // Both channels
+				{
+					for (f_cnt_t frame = 0; frame < in.frames; ++frame)
+					{
+						outPtr[frame] += convertSample<SampleT>(inPtr[frame].left() + inPtr[frame].right());
+					}
+					numRouted += 2;
+					break;
+				}
+				default:
+					unreachable();
+					break;
+			}
+		}
+
+		// Either no input channels were routed to this output and output stays zeroed,
+		// or only one channel was routed and normalization is not needed
+		if (numRouted <= 1) { continue; }
+
+		// Normalize output
+		for (f_cnt_t frame = 0; frame < in.frames; ++frame)
+		{
+			outPtr[frame] /= numRouted;
+		}
+	}
+}
+
+template<AudioDataLayout layout, typename SampleT, int channelCountIn, int channelCountOut>
+inline void PluginPinConnector::Router<layout, SampleT, channelCountIn, channelCountOut>::routeFromPlugin(
+	SplitAudioData<const SampleT, channelCountOut> in, CoreAudioBusMut inOut) const
+{
+	if constexpr (channelCountOut == 0) { return; }
+
+	assert(m_pc->m_out.channelCount() != DynamicChannelCount);
+	if (m_pc->m_out.channelCount() == 0) { return; }
+
+	// Ignore all unused track channels for better performance
+	const auto inOutSizeConstrained = m_pc->m_trackChannelsUpperBound / 2;
+	assert(inOutSizeConstrained <= inOut.channelPairs);
+
+	/*
+	 * Routes plugin audio to track channel pair and normalizes the result. For track channels
+	 * without any plugin audio routed to it, the track channel is unmodified for "bypass"
+	 * behavior.
+	 */
+	const auto routeNx2 = [&](SampleFrame* outPtr, ch_cnt_t outChannel, auto routedChannels) {
+		constexpr std::uint8_t rc = routedChannels();
+
+		if constexpr (rc == 0b00)
+		{
+			// Both track channels bypassed - nothing to do
+			return;
+		}
+
+		// We know at this point that we are writing to at least one of the output channels
+		// rather than bypassing, so it is safe to set the output buffer of those channels
+		// to zero prior to accumulation
+
+		if constexpr (rc == 0b11)
+		{
+			// TODO: std::memcpy?
```

Comment:
The comment is here because I wasn't sure how well the compiler would optimize `std::fill_n` vs `std::memcpy`.

This case where both channels need to be zeroed (that is, when `rc == 0b11`) is handled separately not because it needs to be, but because it's an opportunity for optimization. Zeroing every sample in an array with `memcpy` can be done more efficiently than zeroing every other sample such as when only the left channel or only the right channel needs to be zeroed.